prboom2/src/m_file.c

```diff
@@ -65,22 +70,238 @@
 
 #define MKDIR_NO_ERROR 0
 
+#ifdef _WIN32
+static wchar_t *ConvertMultiByteToWide(const char *str, UINT code_page) {
+  wchar_t *wstr = NULL;
+  int wlen = 0;
+
+  wlen = MultiByteToWideChar(code_page, 0, str, -1, NULL, 0);
+
+  if (!wlen) {
+    errno = EINVAL;
+    lprintf(LO_INFO, "Warning: Failed to convert path to wide encoding\n");
+    return NULL;
+  }
+
+  wstr = malloc(sizeof(wchar_t) * wlen);
+
+  if (!wstr) {
+    lprintf(LO_INFO, "ConvertMultiByteToWide: Failed to allocate new string\n");
+    return NULL;
+  }
+
+  if (MultiByteToWideChar(code_page, 0, str, -1, wstr, wlen) == 0) {
+    errno = EINVAL;
+    lprintf(LO_INFO, "Warning: Failed to convert path to wide encoding\n");
+    free(wstr);
+    return NULL;
+  }
+
+  return wstr;
+}
+
+static char *ConvertWideToMultiByte(const wchar_t *wstr, UINT code_page) {
+  char *str = NULL;
+  int len = 0;
+
+  len = WideCharToMultiByte(code_page, 0, wstr, -1, NULL, 0, NULL, NULL);
+
+  if (!len) {
+    errno = EINVAL;
+    lprintf(LO_INFO, "Warning: Failed to convert path to multi byte encoding\n");
+    return NULL;
+  }
+
+  str = malloc(sizeof(char) * len);
+
+  if (!str) {
+    lprintf(LO_INFO, "ConvertWideToMultiByte: Failed to allocate new string\n");
+    return NULL;
+  }
+
+  if (WideCharToMultiByte(code_page, 0, wstr, -1, str, len, NULL, NULL) == 0) {
+    errno = EINVAL;
+    lprintf(LO_INFO, "Warning: Failed to convert path to multi byte encoding\n");
+    free(str);
+    return NULL;
+  }
+
+  return str;
+}
+
+wchar_t *ConvertUtf8ToWide(const char *str) {
+  return ConvertMultiByteToWide(str, CP_UTF8);
+}
+
+static char *ConvertWideToUtf8(const wchar_t *wstr) {
+  return ConvertWideToMultiByte(wstr, CP_UTF8);
+}
+
+static wchar_t *ConvertSysNativeMBToWide(const char *str) {
+  return ConvertMultiByteToWide(str, CP_ACP);
+}
+
+static char *ConvertWideToSysNativeMB(const wchar_t *wstr) {
+  return ConvertWideToMultiByte(wstr, CP_ACP);
+}
+#endif
+
+char *ConvertSysNativeMBToUtf8(const char *str) {
+#ifdef _WIN32
+  char *ret = NULL;
+  wchar_t *wstr = NULL;
+
+  wstr = ConvertSysNativeMBToWide(str);
+
+  if (!wstr)
+    return NULL;
+
+  ret = ConvertWideToUtf8(wstr);
+
+  free(wstr);
```

Comment:
Could you use zone functions for all the frees and allocations?

Reply:
Then we also need to use the zone memory in `i_glob.c` Is this necessary?

Reply:
Yes, otherwise we'll be using mixed memory management, which sounds like a recipe for future bugs 😄

Reply:
Fixed in https://github.com/kraflab/dsda-doom/pull/234/commits/db2a68cf318320ce2ab36805d8645cf3040967d9